### PR TITLE
Rewrite the device key query logic

### PR DESCRIPTION
**Pull request by richvdh**
Only permit one query per user at a time.

Builds on top of https://github.com/matrix-org/matrix-js-sdk/pull/346.

### CODE REVIEW

File: src/crypto/DeviceList.js [hunk 1]
```diff
@@ -0,0 +1,421 @@
+/*
+Copyright 2017 OpenMarket Ltd
```
Comment (ara4n) [inline src/crypto/DeviceList.js]: VCL

File: src/crypto/DeviceList.js [hunk 2]
```diff
@@ -210,134 +195,124 @@ export default class DeviceList {
      * @param {String} userId
      */
     invalidateUserDeviceList(userId) {
+        if (typeof userId !== 'string') {
```
Comment (ara4n) [inline src/crypto/DeviceList.js]: comment why this guard is needed, or if it's just paranoia?

File: src/crypto/DeviceList.js [hunk 3]
```diff
@@ -210,134 +195,124 @@ export default class DeviceList {
...
 
     /**
-     * Start device queries for any users who sent us an m.new_device recently
+     * Start device queries for any users with outdated device lists
+     *
+     * We tolerate multiple concurrent device queries, but only one query per
+     * user.
+     *
+     * If any users already have downloads in progress, they are ignored -
+     * they will be refreshed when the current download completes anyway.
```
Comment (ara4n) [inline src/crypto/DeviceList.js]: might be worth explaining why the fact this sounds racey (with the downloads completing potentially at quite different times) isn't a problem...

File: src/crypto/DeviceList.js [hunk 4]
```diff
@@ -53,45 +53,30 @@ export default class DeviceList {
...
             }
-        }
+        });
 
-        if (downloadUsers.length > 0) {
-            const r = this._doKeyDownloadForUsers(downloadUsers);
-            downloadUsers.map(function(u) {
-                promises.push(r[u].catch(perUserCatch(u)));
-            });
+        if (needsFlush) {
```
Comment (ara4n) [inline src/crypto/DeviceList.js]: every time i've read the flush code i've been confused, as flush to me always sounds as if you're trying to write to something and you want the writes to all complete.  But here, it's actually flushing *reads* to download new device lists, aiui?  I wonder if there is better naming to be had here...

**Comment by ara4n** [comment]
LGTM, modulo a few q's

File: src/crypto/DeviceList.js (hunk 2 above)
Comment (richvdh) [inline src/crypto/DeviceList.js]: Mostly paranoia. Added a comment.

File: src/crypto/DeviceList.js (hunk 3 above)
Comment (richvdh) [inline src/crypto/DeviceList.js]: Hrm. I've expanded the comment abit, but perhaps I'm missing the source of confusion here?

File: src/crypto/DeviceList.js (hunk 4 above)
Comment (richvdh) [inline src/crypto/DeviceList.js]: Fair point. Renamed to `flushDeviceListRequests`.

**Pull request merged**